Origin: upstream, 7eb5fdd4536019ac8afe50a23e96d9e56aa9c82d
From: Hans de Goede <hdegoede@redhat.com>
Date: Tue, 2 Aug 2011 13:30:23 +0200
Subject: [PATCH] libv4l: Add a force tinyjpeg quirk for w996Xcf based cams

These cams produce non standard JPEG data which our embedded tinyjpeg
copy has been patched to handle, but regular libjpeg cannot handle,
so force the use of our embedded tinyjpeg for these cams.

Signed-off-by: Hans de Goede <hdegoede@redhat.com>
---
 lib/libv4lconvert/control/libv4lcontrol.c |    3 +++
 lib/libv4lconvert/control/libv4lcontrol.h |    1 +
 lib/libv4lconvert/libv4lconvert.c         |    4 +++-
 3 files changed, 7 insertions(+), 1 deletions(-)

Index: libv4l/lib/libv4lconvert/control/libv4lcontrol.c
===================================================================
--- libv4l.orig/lib/libv4lconvert/control/libv4lcontrol.c	2011-08-05 14:35:43.592828552 +0200
+++ libv4l/lib/libv4lconvert/control/libv4lcontrol.c	2011-08-05 14:42:25.716824717 +0200
@@ -633,6 +633,9 @@
 		V4LCONTROL_HFLIPPED | V4LCONTROL_WANTS_WB_AUTOGAIN },
 	{ 0x047d, 0x5003, 0,    NULL, NULL,
 		V4LCONTROL_HFLIPPED | V4LCONTROL_WANTS_WB_AUTOGAIN },
+	/* W996xCF based cams, must use jpeglite because of funky JPEG fmt */
+	{ 0x041e, 0x4003, 0,    NULL, NULL, V4LCONTROL_FORCE_TINYJPEG },
+	{ 0x1046, 0x9967, 0,    NULL, NULL, V4LCONTROL_FORCE_TINYJPEG },
 };
 
 static const struct v4l2_queryctrl fake_controls[];
Index: libv4l/lib/libv4lconvert/control/libv4lcontrol.h
===================================================================
--- libv4l.orig/lib/libv4lconvert/control/libv4lcontrol.h	2011-07-16 15:30:42.772000002 +0200
+++ libv4l/lib/libv4lconvert/control/libv4lcontrol.h	2011-08-05 14:42:25.716824717 +0200
@@ -28,6 +28,7 @@
 #define V4LCONTROL_ROTATED_90_JPEG       0x04
 #define V4LCONTROL_WANTS_WB              0x08
 #define V4LCONTROL_WANTS_AUTOGAIN        0x10
+#define V4LCONTROL_FORCE_TINYJPEG        0x20
 
 /* Masks */
 #define V4LCONTROL_WANTS_WB_AUTOGAIN     (V4LCONTROL_WANTS_WB | V4LCONTROL_WANTS_AUTOGAIN)
Index: libv4l/lib/libv4lconvert/libv4lconvert.c
===================================================================
--- libv4l.orig/lib/libv4lconvert/libv4lconvert.c	2011-08-05 14:35:43.592828552 +0200
+++ libv4l/lib/libv4lconvert/libv4lconvert.c	2011-08-05 14:42:25.716824717 +0200
@@ -160,8 +160,10 @@
 		free(data);
 		return NULL;
 	}
-	data->control_flags = v4lcontrol_get_flags(data->control);
 	data->bandwidth = v4lcontrol_get_bandwidth(data->control);
+	data->control_flags = v4lcontrol_get_flags(data->control);
+	if (data->control_flags & V4LCONTROL_FORCE_TINYJPEG)
+		data->flags |= V4LCONVERT_USE_TINYJPEG;
 
 	data->processing = v4lprocessing_create(fd, data->control);
 	if (!data->processing) {
